src/zcash/History.cpp

```diff
@@ -0,0 +1,120 @@
+#include "zcash/History.hpp"
+
+#include <stdexcept>
+
+#include <boost/foreach.hpp>
+
+#include "serialize.h"
+#include "streams.h"
+#include "uint256.h"
+#include "librustzcash.h"
+
+namespace libzcash {
+
+void HistoryCache::Extend(const HistoryNode &leaf) {
+    appends[length++] = leaf;
+}
+
+void HistoryCache::Truncate(HistoryIndex newLength) {
+    for (HistoryIndex idx = length; idx > newLength; idx--) {
+        appends.erase(idx);
+    }
+
+    length = newLength;
+    // we track how deep updates go back in the tree, so we could later
+    // update everything starting from `updateDepth`
+    //
+    // imagine we rolled two blocks back and then put another 3 blocks on top
+    // of the rolled back state. In that case `updateDepth` will be H-3, while length
+    // will be H (where H is a final chain height after such operation). So we know that
+    // history entries in the range of H-3..H are expected to be pushed into the database
+    // to replace/append to the persistent nodes there.
+    if (updateDepth > length) updateDepth = length;
+}
+
+HistoryNode NewNode(
+        uint256 subtreeCommitment,
+        uint32_t startTime,
+        uint32_t endTime,
+        uint32_t startTarget,
+        uint32_t endTarget,
+        uint256 startSaplingRoot,
+        uint256 endSaplingRoot,
+        uint256 subtreeTotalWork,
+        uint64_t startHeight,
+        uint64_t endHeight,
+        uint64_t saplingTxCount
+    )
+{
+    CDataStream buf(SER_DISK, 0);
+    HistoryNode result;
+
+    buf << subtreeCommitment;
+    buf << startTime;
+    buf << endTime;
+    buf << startTarget;
+    buf << endTarget;
+    buf << startSaplingRoot;
+    buf << endSaplingRoot;
+    buf << subtreeTotalWork;
+    buf << COMPACTSIZE(startHeight);
+    buf << COMPACTSIZE(endHeight);
+    buf << COMPACTSIZE(saplingTxCount);
+
+    std::copy(buf.begin(), buf.end(), result.begin());
+    return result;
+}
+
+HistoryNode NewLeaf(
+    uint256 commitment,
+    uint32_t time,
+    uint32_t target,
+    uint256 saplingRoot,
+    uint256 totalWork,
+    uint64_t height,
+    uint64_t saplingTxCount
+) {
+    return NewNode(
+        commitment,
+        time,
+        time,
+        target,
+        target,
+        saplingRoot,
+        saplingRoot,
+        totalWork,
+        height,
+        height,
+        saplingTxCount
+    );
+}
+
+HistoryEntry NodeToEntry(const HistoryNode node, uint32_t left, uint32_t right) {
+    CDataStream buf(SER_DISK, 0);
+    HistoryEntry result;
+
+    uint8_t code = 0;
+    buf << code;
+    buf << left;
+    buf << right;
+    buf << node;
+
+    std::copy(std::begin(buf), std::end(buf), std::begin(result));
```

Comment:
Could have assertion here (and similarly in `LeafToEntry`) that the `CDataStream` buffer size is equal to `NODE_SERIALIZED_LENGTH`.

Reply:
I think this point still stands after the later commits.

Reply:
It won't be equal, because the last three fields are `CompactSize` fields. I could assert that the buffer size is not greater than `ENTRY_SERIALIZED_LENGTH` in this particular case.